diff --git a/sdk/BUILD.gn b/sdk/BUILD.gn
index 839549ac95..57380f8a43 100644
--- a/sdk/BUILD.gn
+++ b/sdk/BUILD.gn
@@ -141,27 +141,34 @@ if (is_ios || is_mac) {
 
   rtc_library("helpers_objc") {
     sources = [
-      "objc/helpers/AVCaptureSession+DevicePosition.h",
-      "objc/helpers/AVCaptureSession+DevicePosition.mm",
       "objc/helpers/NSString+StdString.h",
       "objc/helpers/NSString+StdString.mm",
       "objc/helpers/RTCDispatcher+Private.h",
       "objc/helpers/RTCDispatcher.h",
       "objc/helpers/RTCDispatcher.m",
-      "objc/helpers/scoped_cftyperef.h",
     ]
 
     deps = [
-      ":base_objc",
-      "../rtc_base:checks",
       "//third_party/abseil-cpp/absl/strings:string_view",
     ]
 
-    frameworks = [
-      "AVFoundation.framework",
-      "CoreMedia.framework",
-      "Foundation.framework",
-    ]
+    frameworks = [ "Foundation.framework" ]
+
+    if (!rtc_data_channel_only) {
+      sources += [
+        "objc/helpers/AVCaptureSession+DevicePosition.h",
+        "objc/helpers/AVCaptureSession+DevicePosition.mm",
+        "objc/helpers/scoped_cftyperef.h",
+      ]
+      deps += [
+        ":base_objc",
+        "../rtc_base:checks",
+      ]
+      frameworks += [
+        "AVFoundation.framework",
+        "CoreMedia.framework",
+      ]
+    }
 
     configs += [
       "..:common_objc",
@@ -171,19 +178,21 @@ if (is_ios || is_mac) {
     public_configs = [ ":common_config_objc" ]
 
     if (is_ios) {
-      sources += [
-        "objc/helpers/UIDevice+RTCDevice.h",
-        "objc/helpers/UIDevice+RTCDevice.mm",
-      ]
-
-      if (target_platform != "tvos") {
+      if (!rtc_data_channel_only) {
         sources += [
-          "objc/helpers/RTCCameraPreviewView.h",
-          "objc/helpers/RTCCameraPreviewView.m",
+          "objc/helpers/UIDevice+RTCDevice.h",
+          "objc/helpers/UIDevice+RTCDevice.mm",
         ]
-      }
 
-      frameworks += [ "UIKit.framework" ]
+        if (target_platform != "tvos") {
+          sources += [
+            "objc/helpers/RTCCameraPreviewView.h",
+            "objc/helpers/RTCCameraPreviewView.m",
+          ]
+        }
+
+        frameworks += [ "UIKit.framework" ]
+      }
     }
   }
 
@@ -971,6 +980,10 @@ if (is_ios || is_mac) {
         "environment_construction",
       ]
       configs += [ "..:no_global_constructors" ]
+      if (rtc_data_channel_only) {
+        defines = [ "HAVE_NO_MEDIA" ]
+        cflags = [ "-Wno-incomplete-implementation" ]
+      }
       sources = [
         "objc/api/peerconnection/RTCAudioSource+Private.h",
         "objc/api/peerconnection/RTCAudioSource.h",
@@ -1088,6 +1101,39 @@ if (is_ios || is_mac) {
         "objc/api/peerconnection/RTCVideoTrack.h",
         "objc/api/peerconnection/RTCVideoTrack.mm",
       ]
+      if (rtc_data_channel_only) {
+        sources -= [
+          "objc/api/peerconnection/RTCAudioSource.mm",
+          "objc/api/peerconnection/RTCAudioTrack.mm",
+          "objc/api/peerconnection/RTCFieldTrials.mm",
+          "objc/api/peerconnection/RTCDtmfSender.mm",
+          "objc/api/peerconnection/RTCLegacyStatsReport.mm",
+          "objc/api/peerconnection/RTCMediaStream.mm",
+          "objc/api/peerconnection/RTCMediaStreamTrack.mm",
+          "objc/api/peerconnection/RTCMetrics.mm",
+          "objc/api/peerconnection/RTCMetricsSampleInfo.mm",
+          "objc/api/peerconnection/RTCPeerConnection+Stats.mm",
+          "objc/api/peerconnection/RTCPeerConnectionFactoryBuilder+DefaultComponents.mm",
+          "objc/api/peerconnection/RTCPeerConnectionFactoryBuilder.mm",
+          "objc/api/peerconnection/RTCPeerConnectionFactoryOptions.mm",
+          "objc/api/peerconnection/RTCRtcpParameters.mm",
+          "objc/api/peerconnection/RTCRtpCapabilities.mm",
+          "objc/api/peerconnection/RTCRtpCodecCapability.mm",
+          "objc/api/peerconnection/RTCRtpCodecParameters.mm",
+          "objc/api/peerconnection/RTCRtpEncodingParameters.mm",
+          "objc/api/peerconnection/RTCRtpHeaderExtension.mm",
+          "objc/api/peerconnection/RTCRtpHeaderExtensionCapability.mm",
+          "objc/api/peerconnection/RTCRtpParameters.mm",
+          "objc/api/peerconnection/RTCRtpReceiver.mm",
+          "objc/api/peerconnection/RTCRtpSender.mm",
+          "objc/api/peerconnection/RTCRtpSource.mm",
+          "objc/api/peerconnection/RTCRtpTransceiver.mm",
+          "objc/api/peerconnection/RTCSSLAdapter.mm",
+          "objc/api/peerconnection/RTCStatisticsReport.mm",
+          "objc/api/peerconnection/RTCTracing.mm",
+          "objc/api/peerconnection/RTCVideoTrack.mm",
+        ]
+      }
 
       configs += [
         "..:common_objc",
@@ -1173,6 +1219,38 @@ if (is_ios || is_mac) {
           "../api/transport:network_control",
         ]
       }
+      if (rtc_data_channel_only) {
+        deps -= [
+          ":audio_device_api_objc",
+          ":base_native_additions_objc",
+          ":native_api_audio_device_module",
+          ":native_video",
+          ":objc_audio_device_module",
+          ":videoframebuffer_objc",
+          ":videorendereradapter_objc",
+          ":videosource_objc",
+          ":videotoolbox_objc",
+          "../api:dtmf_sender_interface",
+          "../api:enable_media",
+          "../api/audio:audio_device",
+          "../api/audio:audio_processing",
+          "../api/audio:builtin_audio_processing_builder",
+          "../api/audio_codecs:audio_codecs_api",
+          "../api/audio_codecs:builtin_audio_decoder_factory",
+          "../api/audio_codecs:builtin_audio_encoder_factory",
+          "../api/crypto:frame_decryptor_interface",
+          "../api/crypto:frame_encryptor_interface",
+          "../api/rtc_event_log:rtc_event_log_factory",
+          "../api/task_queue:default_task_queue_factory",
+          "../api/video:video_frame",
+          "../api/video:video_rtp_headers",
+          "../api/video_codecs:video_codecs_api",
+          "../common_video",
+          "../media:media_constants",
+          "../media:rtc_media_base",
+          "../modules/video_coding:video_codec_interface",
+        ]
+      }
     }
 
     if (rtc_include_tests) {
@@ -1435,23 +1513,43 @@ if (is_ios || is_mac) {
         public_headers = common_objc_headers
 
         ldflags = [
-          "-all_load",
           "-install_name",
           "@rpath/$output_name.framework/$output_name",
         ]
+        if (rtc_data_channel_only) {
+          ldflags += [
+            "-force_load",
+            rebase_path(
+                "$root_out_dir/obj/sdk/libpeerconnectionfactory_base_objc.a",
+                root_build_dir),
+            "-force_load",
+            rebase_path(
+                "$root_out_dir/obj/sdk/libmediaconstraints_objc.a",
+                root_build_dir),
+            "-force_load",
+            rebase_path("$root_out_dir/obj/sdk/libhelpers_objc.a",
+                        root_build_dir),
+          ]
+        } else {
+          ldflags += [ "-all_load" ]
+        }
 
         deps = [
-          ":audio_objc",
           ":base_objc",
-          ":default_codec_factory_objc",
-          ":metal_objc",
           ":native_api",
-          ":native_video",
           ":peerconnectionfactory_base_objc",
-          ":videocapture_objc",
-          ":videocodec_objc",
-          ":videotoolbox_objc",
         ]
+        if (!rtc_data_channel_only) {
+          deps += [
+            ":audio_objc",
+            ":default_codec_factory_objc",
+            ":metal_objc",
+            ":native_video",
+            ":videocapture_objc",
+            ":videocodec_objc",
+            ":videotoolbox_objc",
+          ]
+        }
         if (!build_with_chromium) {
           deps += [
             ":callback_logger_objc",
@@ -1689,6 +1787,36 @@ if (is_ios || is_mac) {
       if (is_ios) {
         deps += [ ":native_network_monitor" ]
       }
+      if (rtc_data_channel_only) {
+        sources -= [
+          "objc/native/api/video_capturer.h",
+          "objc/native/api/video_capturer.mm",
+          "objc/native/api/video_decoder_factory.h",
+          "objc/native/api/video_decoder_factory.mm",
+          "objc/native/api/video_encoder_factory.h",
+          "objc/native/api/video_encoder_factory.mm",
+          "objc/native/api/video_frame.h",
+          "objc/native/api/video_frame.mm",
+          "objc/native/api/video_frame_buffer.h",
+          "objc/native/api/video_frame_buffer.mm",
+          "objc/native/api/video_renderer.h",
+          "objc/native/api/video_renderer.mm",
+        ]
+        deps -= [
+          ":native_video",
+          ":videoframebuffer_objc",
+          "../api:libjingle_peerconnection_api",
+          "../api:make_ref_counted",
+          "../api:media_stream_interface",
+          "../api:scoped_refptr",
+          "../api/environment",
+          "../api/video:video_frame",
+          "../api/video:video_rtp_headers",
+          "../api/video_codecs:video_codecs_api",
+          "../common_video",
+          "../pc:video_track_source_proxy",
+        ]
+      }
     }
 
     if (is_ios) {
diff --git a/sdk/objc/api/peerconnection/RTCPeerConnection.mm b/sdk/objc/api/peerconnection/RTCPeerConnection.mm
index fb09754751..c2ea0fb7ad 100644
--- a/sdk/objc/api/peerconnection/RTCPeerConnection.mm
+++ b/sdk/objc/api/peerconnection/RTCPeerConnection.mm
@@ -14,14 +14,17 @@
 #import "RTCDataChannel+Private.h"
 #import "RTCIceCandidate+Private.h"
 #import "RTCIceCandidateErrorEvent+Private.h"
-#import "RTCLegacyStatsReport+Private.h"
 #import "RTCMediaConstraints+Private.h"
+#ifndef HAVE_NO_MEDIA
 #import "RTCMediaStream+Private.h"
 #import "RTCMediaStreamTrack+Private.h"
+#endif
 #import "RTCPeerConnectionFactory+Private.h"
+#ifndef HAVE_NO_MEDIA
 #import "RTCRtpReceiver+Private.h"
 #import "RTCRtpSender+Private.h"
 #import "RTCRtpTransceiver+Private.h"
+#endif
 #import "RTCSessionDescription+Private.h"
 #import "base/RTCLogging.h"
 #import "helpers/NSString+StdString.h"
@@ -151,6 +154,9 @@ void OnFailure(RTCError error) override {
 
 void PeerConnectionDelegateAdapter::OnAddStream(
     webrtc::scoped_refptr<MediaStreamInterface> stream) {
+#ifdef HAVE_NO_MEDIA
+  (void)stream;
+#else
   RTC_OBJC_TYPE(RTCPeerConnection) *peer_connection = peer_connection_;
   if (peer_connection == nil) {
     return;
@@ -165,10 +171,14 @@ void OnFailure(RTCError error) override {
       alloc] initWithFactory:peer_connection.factory nativeMediaStream:stream];
 
   [delegate peerConnection:peer_connection didAddStream:media_stream];
+#endif
 }
 
 void PeerConnectionDelegateAdapter::OnRemoveStream(
     webrtc::scoped_refptr<MediaStreamInterface> stream) {
+#ifdef HAVE_NO_MEDIA
+  (void)stream;
+#else
   RTC_OBJC_TYPE(RTCPeerConnection) *peer_connection = peer_connection_;
   if (peer_connection == nil) {
     return;
@@ -182,10 +192,14 @@ void OnFailure(RTCError error) override {
       alloc] initWithFactory:peer_connection.factory nativeMediaStream:stream];
 
   [delegate peerConnection:peer_connection didRemoveStream:mediaStream];
+#endif
 }
 
 void PeerConnectionDelegateAdapter::OnTrack(
     webrtc::scoped_refptr<RtpTransceiverInterface> nativeTransceiver) {
+#ifdef HAVE_NO_MEDIA
+  (void)nativeTransceiver;
+#else
   RTC_OBJC_TYPE(RTCPeerConnection) *peer_connection = peer_connection_;
   if (peer_connection == nil) {
     return;
@@ -205,6 +219,7 @@ void OnFailure(RTCError error) override {
     [delegate peerConnection:peer_connection
         didStartReceivingOnTransceiver:transceiver];
   }
+#endif
 }
 
 void PeerConnectionDelegateAdapter::OnDataChannel(
@@ -411,6 +426,10 @@ IceCandidate remote_candidate_wrapper(
 void PeerConnectionDelegateAdapter::OnAddTrack(
     webrtc::scoped_refptr<RtpReceiverInterface> receiver,
     const std::vector<webrtc::scoped_refptr<MediaStreamInterface>> &streams) {
+#ifdef HAVE_NO_MEDIA
+  (void)receiver;
+  (void)streams;
+#else
   RTC_OBJC_TYPE(RTCPeerConnection) *peer_connection = peer_connection_;
   if (peer_connection == nil) {
     return;
@@ -440,10 +459,14 @@ IceCandidate remote_candidate_wrapper(
               didAddReceiver:rtpReceiver
                      streams:mediaStreams];
   }
+#endif
 }
 
 void PeerConnectionDelegateAdapter::OnRemoveTrack(
     webrtc::scoped_refptr<RtpReceiverInterface> receiver) {
+#ifdef HAVE_NO_MEDIA
+  (void)receiver;
+#else
   RTC_OBJC_TYPE(RTCPeerConnection) *peer_connection = peer_connection_;
   if (peer_connection == nil) {
     return;
@@ -460,13 +483,16 @@ IceCandidate remote_candidate_wrapper(
              nativeRtpReceiver:receiver];
     [delegate peerConnection:peer_connection didRemoveReceiver:rtpReceiver];
   }
+#endif
 }
 
 }  // namespace webrtc
 
 @implementation RTC_OBJC_TYPE (RTCPeerConnection) {
   RTC_OBJC_TYPE(RTCPeerConnectionFactory) * _factory;
+#ifndef HAVE_NO_MEDIA
   NSMutableArray<RTC_OBJC_TYPE(RTCMediaStream) *> *_localStreams;
+#endif
   std::unique_ptr<webrtc::PeerConnectionDelegateAdapter> _observer;
   webrtc::scoped_refptr<webrtc::PeerConnectionInterface> _peerConnection;
   std::unique_ptr<webrtc::MediaConstraints> _nativeConstraints;
@@ -528,15 +554,19 @@ @implementation RTC_OBJC_TYPE (RTCPeerConnection) {
     }
     _peerConnection = result.MoveValue();
     _factory = factory;
+#ifndef HAVE_NO_MEDIA
     _localStreams = [[NSMutableArray alloc] init];
+#endif
     _delegate = delegate;
   }
   return self;
 }
 
+#ifndef HAVE_NO_MEDIA
 - (NSArray<RTC_OBJC_TYPE(RTCMediaStream) *> *)localStreams {
   return [_localStreams copy];
 }
+#endif
 
 - (RTC_OBJC_TYPE(RTCSessionDescription) *)localDescription {
   // It's only safe to operate on SessionDescriptionInterface on the signaling
@@ -636,6 +666,7 @@ - (void)removeIceCandidates:
   }
 }
 
+#ifndef HAVE_NO_MEDIA
 - (void)addStream:(RTC_OBJC_TYPE(RTCMediaStream) *)stream {
   if (!_peerConnection->AddStream(stream.nativeMediaStream.get())) {
     RTCLogError(@"Failed to add stream: %@", stream);
@@ -725,8 +756,9 @@ - (BOOL)removeTrack:(RTC_OBJC_TYPE(RTCRtpSender) *)sender {
   }
   return [[RTC_OBJC_TYPE(RTCRtpTransceiver) alloc]
            initWithFactory:self.factory
-      nativeRtpTransceiver:nativeTransceiverOrError.MoveValue()];
+            nativeRtpTransceiver:nativeTransceiverOrError.MoveValue()];
 }
+#endif
 
 - (void)restartIce {
   _peerConnection->RestartIce();
@@ -834,6 +866,7 @@ - (void)stopRtcEventLog {
   _hasStartedRtcEventLog = NO;
 }
 
+#ifndef HAVE_NO_MEDIA
 - (RTC_OBJC_TYPE(RTCRtpSender) *)senderWithKind:(NSString *)kind
                                        streamId:(NSString *)streamId {
   std::string nativeKind = [NSString stdStringForString:kind];
@@ -885,6 +918,7 @@ - (void)stopRtcEventLog {
   }
   return transceivers;
 }
+#endif
 
 #pragma mark - Private
 
diff --git a/sdk/objc/api/peerconnection/RTCPeerConnectionFactory.mm b/sdk/objc/api/peerconnection/RTCPeerConnectionFactory.mm
index 2ea5393960..3023a6c298 100644
--- a/sdk/objc/api/peerconnection/RTCPeerConnectionFactory.mm
+++ b/sdk/objc/api/peerconnection/RTCPeerConnectionFactory.mm
@@ -14,35 +14,38 @@
 #import "RTCPeerConnectionFactory+Native.h"
 #import "RTCPeerConnectionFactory+Private.h"
 #import "RTCPeerConnectionFactoryOptions+Private.h"
-#import "RTCRtpCapabilities+Private.h"
 
+#ifndef HAVE_NO_MEDIA
+#import "RTCRtpCapabilities+Private.h"
 #import "RTCAudioSource+Private.h"
 #import "RTCAudioTrack+Private.h"
-#import "RTCMediaConstraints+Private.h"
 #import "RTCMediaStream+Private.h"
-#import "RTCPeerConnection+Private.h"
 #import "RTCVideoSource+Private.h"
 #import "RTCVideoTrack+Private.h"
-#import "base/RTCLogging.h"
 #import "base/RTCVideoDecoderFactory.h"
 #import "base/RTCVideoEncoderFactory.h"
+#endif
+#import "RTCMediaConstraints+Private.h"
+#import "RTCPeerConnection+Private.h"
+#import "base/RTCLogging.h"
 #import "helpers/NSString+StdString.h"
 #include "rtc_base/checks.h"
 #include "sdk/objc/native/api/network_monitor_factory.h"
 #include "sdk/objc/native/api/ssl_certificate_verifier.h"
 
+#include "api/create_modular_peer_connection_factory.h"
+#include "api/environment/environment.h"
+#include "api/environment/environment_factory.h"
+#include "api/scoped_refptr.h"
+#ifndef HAVE_NO_MEDIA
 #include "api/audio/audio_device.h"
 #include "api/audio/audio_processing.h"
 #include "api/audio/builtin_audio_processing_builder.h"
 #include "api/audio_codecs/builtin_audio_decoder_factory.h"
 #include "api/audio_codecs/builtin_audio_encoder_factory.h"
-#include "api/create_modular_peer_connection_factory.h"
 #include "api/enable_media.h"
-#include "api/environment/environment.h"
-#include "api/environment/environment_factory.h"
 #include "api/media_stream_interface.h"
 #include "api/rtc_event_log/rtc_event_log_factory.h"
-#include "api/scoped_refptr.h"
 #import "components/video_codec/RTCVideoDecoderFactoryH264.h"
 #import "components/video_codec/RTCVideoEncoderFactoryH264.h"
 #include "media/base/media_constants.h"
@@ -57,6 +60,7 @@
 #if defined(WEBRTC_IOS)
 #import "sdk/objc/native/api/audio_device_module.h"
 #endif
+#endif
 
 @implementation RTC_OBJC_TYPE (RTCPeerConnectionFactory) {
   std::optional<webrtc::Environment> _env;
@@ -69,6 +73,9 @@ @implementation RTC_OBJC_TYPE (RTCPeerConnectionFactory) {
 @synthesize nativeFactory = _nativeFactory;
 
 - (instancetype)init {
+#ifdef HAVE_NO_MEDIA
+  return [self initWithNoMedia];
+#else
   webrtc::PeerConnectionFactoryDependencies dependencies;
   dependencies.audio_encoder_factory =
       webrtc::CreateBuiltinAudioEncoderFactory();
@@ -83,6 +90,7 @@ @implementation RTC_OBJC_TYPE (RTCPeerConnectionFactory) {
   dependencies.adm = webrtc::CreateAudioDeviceModule(*dependencies.env);
 #endif
   return [self initWithMediaAndDependencies:dependencies];
+#endif
 }
 
 - (instancetype)
@@ -177,6 +185,7 @@ @implementation RTC_OBJC_TYPE (RTCPeerConnectionFactory) {
   return [self initWithNativeDependencies:default_deps];
 }
 
+#ifndef HAVE_NO_MEDIA
 - (instancetype)
     initWithNativeAudioEncoderFactory:
         (webrtc::scoped_refptr<webrtc::AudioEncoderFactory>)audioEncoderFactory
@@ -333,6 +342,7 @@ @implementation RTC_OBJC_TYPE (RTCPeerConnectionFactory) {
   return [[RTC_OBJC_TYPE(RTCMediaStream) alloc] initWithFactory:self
                                                        streamId:streamId];
 }
+#endif
 
 - (nullable RTC_OBJC_TYPE(RTCPeerConnection) *)
     peerConnectionWithConfiguration:
@@ -429,6 +439,7 @@ @implementation RTC_OBJC_TYPE (RTCPeerConnectionFactory) {
 
 #pragma mark - Private
 
+#ifndef HAVE_NO_MEDIA
 + (webrtc::MediaType)mediaTypeForKind:(NSString *)kind {
   if (kind == kRTCMediaStreamTrackKindAudio) {
     return webrtc::MediaType::AUDIO;
@@ -439,5 +450,6 @@ @implementation RTC_OBJC_TYPE (RTCPeerConnectionFactory) {
     return webrtc::MediaType::UNSUPPORTED;
   }
 }
+#endif
 
 @end
diff --git a/webrtc.gni b/webrtc.gni
index 5c69818f10..26bae493fa 100644
--- a/webrtc.gni
+++ b/webrtc.gni
@@ -47,6 +47,8 @@ if (build_with_chromium) {
 # in this one, can be read from the next one (args defined in the same
 # declare_args cannot be referenced in that scope).
 declare_args() {
+  rtc_data_channel_only = false
+
   # Setting this to true will make RTC_EXPORT (see rtc_base/system/rtc_export.h)
   # expand to code that will manage symbols visibility.
   rtc_enable_symbol_export = false
